src/storage-controller/src/collection_mgmt.rs

```diff
@@ -196,6 +196,20 @@ where
 
                     // Listen for a shutdown signal so we can gracefully cleanup.
                     _ = &mut shutdown_rx => {
+                        let mut senders = Vec::new();
+
+                        // Get as many waiting senders as possible.
+                        'collect: while let Ok((_batch, sender)) = rx.try_recv() {
+                            senders.push(sender);
+
+                            if senders.len() >= CHANNEL_CAPACITY {
+                                break 'collect;
+                            }
+                        }
+
+                        // Notify them that this collection is closed.
+                        notify_listeners(senders, || Err(StorageError::IdentifierInvalid(id)));
```

Comment:
Previously we would return `ShuttingDown` when the senders were dropped which after this PR would return a 503... which is almost correct, but not exactly. The crux of the issue is differentiating between a dropped sender because all of `environmentd` is shutting down (now this returns `ShuttingDown`) and the sender getting dropped because this source was dropped (now this returns `InvalidIdentifier`).

I briefly added a different error variant of `Closed` which we'd return if a source was dropped, but that seemed redundant, so I went with `IdentifierInvalid`. More than happy to add a new error variant here, or change it around, what do you think?

Reply:
Ah ok, maybe add a comment explaining that the source has been dropped which is why we use `IdentifierInvalid`. Also, what happens to the senders that we don't notify in the cases that there are more senders than `CHANNEL_CAPACITY`?